File: spsc-lite-ruby/SLLParser_Tests.rb
require 'rubygems'
require 'test/unit'
require 'rparsec'
require 'rparsec/parsers'
require 'SLL'
require 'SLLParser'

class SLLParser_Tests < Test::Unit::TestCase
  include SLLParser
  def exc(p, input)
    assert_raises(ParserException){(p << Parsers::eof).parse(input)}
  end

  def stringOK(p, expected, input)
    assert_equal(expected, (p << Parsers::eof).parse(input))
  end

  def parseOK(p, expected, input)
    assert_equal(expected, (p << Parsers::eof).parse(input).to_s)
  end

  def expOK(expected, input)
    parseOK(Expression, expected, input)
  end

  def programOK(expected, input)
    parseOK(Program, expected, input)
  end

  def test101Idents()
    stringOK(LIdent, 'aZ12b', 'aZ12b')
    stringOK(UIdent, "Aaa9b", "Aaa9b")
    stringOK(FIdent, "fAa9b", "fAa9b")
    stringOK(GIdent, "gAa9b", "gAa9b")
    exc(LIdent, "Aa")
    exc(UIdent, "aa")
    exc(LIdent, "A*a")
  end

  def test102Call()
    expOK("x", "x")
    expOK("fA", "fA")
    expOK("gA", "gA")
    expOK("A(x,y)", "A(x,y)")
    expOK("A", "A()")
    expOK("A", "A")
    expOK("fA(x,y)", "fA(x,y)")
    expOK("fA()", "fA()")
    expOK("gA(x,y)", "gA(x,y)")
    exc(Expression, "gA()")
  end

  def test103Program()
    programOK("", "")
    programOK("f(x)=A;", "f(x)=A;")
    programOK("f(x)=A;f1(x)=A1;", "f(x)=A;f1(x)=A1;")
    programOK("f()=f();", "f()=f();")
    programOK("g(C)=A;", "g(C)=A;")
    programOK("g(C(x))=A;", "g(C(x))=A;")
    programOK("g(C(x))=A;", "g(C(x))=A;")
    programOK("g(C(x,y),z)=A;", "g(C(x,y),z)=A;")
    programOK("g(C,y)=A;", "g(C,y)=A;")
  end
end

File: spsc-lite-python/src/spsc/sll_parser_tests.py
'''
Created on Aug 14, 2009

@author: Sergei Romanenko
'''

import unittest

from pyparsing import stringEnd, ParseException

from sll_language import *
from sll_parser import *

class SLLParser_Tests(unittest.TestCase):

    def stringOK(self, p, expected, input):
        tokens = p.parseString(input, True)
        self.assertEqual(expected, tokens[0])

    def parseOK(self, p, expected, input):
        tokens = p.parseString(input, True)
        self.assertEqual(expected, "%s" % tokens[0])

    def expOK(self, expected, input):
        self.parseOK(exp, expected, input)

    def programOK(self, expected, input):
        self.parseOK(program, expected, input)

    def exc(self, p, input):
        self.assertRaises(ParseException, p.parseString, input, True)

    def testIdents(self):
        self.stringOK(lIdent, "aAa9b", "aAa9b")
        self.stringOK(uIdent, "Aaa9b", "Aaa9b")
        self.stringOK(fIdent, "fAa9b", "fAa9b")
        self.stringOK(gIdent, "gAa9b", "gAa9b")
        self.exc(lIdent, "Aa")
        self.exc(uIdent, "aa")
        self.exc(lIdent, "A*a")

    def testCall(self):
        self.expOK("x", "x")
        self.expOK("fA", "fA")
        self.expOK("gA", "gA")
        self.expOK("A(x,y)", "A(x,y)")
        self.expOK("A", "A()")
        self.expOK("A", "A")
        self.expOK("fA(x,y)", "fA(x,y)")
        self.expOK("fA()", "fA()")
        self.expOK("gA(x,y)", "gA(x,y)")
        self.exc(exp, "gA()")

    def testProgram(self):
        self.programOK("", "")
        self.programOK("f(x)=A;", "f(x)=A;")
        self.programOK("f(x)=A;f1(x)=A1;", "f(x)=A;f1(x)=A1;")
        self.programOK("f()=f();", "f()=f();")
        self.programOK("g(C)=A;", "g(C)=A;")
        self.programOK("g(C(x))=A;", "g(C(x))=A;")
        self.programOK("g(C(x))=A;", "g(C(x))=A;")
        self.programOK("g(C(x,y),z)=A;", "g(C(x,y),z)=A;")
        self.programOK("g(C,y)=A;", "g(C,y)=A;")


if __name__ == "__main__":
    #import sys;sys.argv = ['', 'Test.testName']
    unittest.main()
